src/Http/Http.Results/src/ResultsOfTHelper.cs

```diff
@@ -1,21 +1,48 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
+using System.Runtime.CompilerServices;
 using Microsoft.AspNetCore.Builder;
 using Microsoft.AspNetCore.Http.Metadata;
 
 namespace Microsoft.AspNetCore.Http;
 
 internal static class ResultsOfTHelper
 {
+    public const DynamicallyAccessedMemberTypes RequireMethods = DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods;
     private static readonly MethodInfo PopulateMetadataMethod = typeof(ResultsOfTHelper).GetMethod(nameof(PopulateMetadata), BindingFlags.Static | BindingFlags.NonPublic)!;
 
-    public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<TTarget>(MethodInfo method, EndpointBuilder builder)
+    public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<[DynamicallyAccessedMembers(RequireMethods)] TTarget>(MethodInfo method, EndpointBuilder builder)
     {
         if (typeof(IEndpointMetadataProvider).IsAssignableFrom(typeof(TTarget)))
         {
-            PopulateMetadataMethod.MakeGenericMethod(typeof(TTarget)).Invoke(null, new object[] { method, builder });
+            var parameters = new object[] { method, builder };
+
+            if (RuntimeFeature.IsDynamicCodeSupported)
+            {
+                InvokeGenericPopulateMetadata(parameters);
+            }
+            else
+            {
+                // Prioritize explicit implementation.
+                var populateMetadataMethod = typeof(TTarget).GetMethod("Microsoft.AspNetCore.Http.Metadata.IEndpointMetadataProvider.PopulateMetadata", BindingFlags.Static | BindingFlags.NonPublic);
```

Comment:
We should write tests for this.... As of now it is untested code, right?

Reply:
Yes. Has the IsDynamicCodeSupported arrived in aspnetcore?

https://github.com/dotnet/aspnetcore/issues/45860

Reply:
Yes, that one is in the current runtime we are using in aspnetcore. It's the DI switch that isn't here yet - blocked on a runtime bug in #46055.

Note even if the IsDynamicCodeSupported wasn't in, we could still write tests by making targeted console apps that are published for AOT and executed.